design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+destined for their tag.  This is how committed data becomes available to
+clients for reading.
+
+Transaction logs internally handle commits via performing two operations
+concurrently.  First, they walk through each mutation in the commit, and push
+the mutation onto an in-memory queue of mutations destined for that tag.
+Second, they include the data in the next batch of pages to durably persist to
+disk.  These queues are popped from when the corresponding storage server has
+persisted the data to its own disk.
+
+TLogs will need to hold the last 5-7 seconds of mutations.  In normal
+operation, the default 1.5GB of memory is enough such that the last 5-7 seconds
+of commits should almost always fit in memory.  However, in the presence of
+failures, the transaction log can be required to buffer significantly more
+data.  Most notably, when a storage server fails, its tag isn't popped until
+data distribution is able to re-replicate all of the shards that storage server
+was responsible for to other storage servers.  Before that happens, mutations
+will accumulate on the TLog destined for the failed storage server, in case it
+comes back and is able to rejoin the cluster.
+
+When this accumulation causes the memory required to hold all the unpopped data
+to exceed `TLOG_SPILL_THREASHOLD` bytes, the transaction log offloads the
+oldest data to disk.  This writing of data to disk to reduce TLog memory
+pressure is referred to as *spilling*.
+
+Previously, spilling would work by writing the data to a SQLite B-tree.  The
+key would be `(tag, version)`, and the value would be all the mutations
+destined for the given tag at the given version.  Peek requests have a start
+version, that is the latest version for which the storage server knows about,
+and the TLog responds by range-reading the B-tree from the start version.  Pop
+requests allow the TLog to forget all mutations for a tag until a specific
+version, and the TLog thus issues a range clear from `(tag, 0)` to
+`(tag, pop_version)`.  After spilling, the durably written data in the disk
+queue would be trimmed to only include from the spilled version on, as any
+required data is now entirely, durably held in the B-tree.  As the entire value
+is copied into the B-tree, this method of spilling will be referred to as
+*spill-by-value* in the rest of this document.
+
+**************************************************************
+*                      Transaction Log                       *
+*                                                            *
+*                                                            *
+*  +------------------+  appends   +------------+            *
+*  | Incoming Commits |----------->| Disk Queue |  +------+  *
+*  +------------------+            +------------+  |SQLite|  *
+*          |                                ^      +------+  *
+*          |                                |        ^       *
+*          |                               pops      |       *
```

Comment:
Ah ok after reading a little further I think I get it now. Basically once data is spilled into the BTree (in the version that contains the full value in the BTree) it then becomes safe to "clean up" the disk queue to remove the data that was spilled since its now persisted in the btree